feature: Layered Navigation (Filters)

codeAdrian opened the pull request:
<!--
Thank you for your contribution!

Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento-research/pwa-studio/blob/master/.github/CONTRIBUTION.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PRs that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR.
-->

## Description
<!--- Provide a general summary of your changes in the Title above -->
<!--- Describe your changes in detail here: -->

When this PR is merged, it will: 
* Add a Layered navigation (Filters) modal where users can choose, apply and reset filters.
* Filters data is pulled from Products data. 
* GraphQL currently doesn't support filters being applied to queries.
* URL is being updated with applied filters
* Applied filters are initialized with URL data
* Config for each filter group (grid/list mode, search, etc.) is partially hard-coded in constants.js file
* Users are able to search the groups of filters (Discussed with @soumya-ashok in meeting). Currently this is applied only to "Fashion color" filters, but it can be enabled for any group.

Additional notes:
* Tests will be added when general components and structure is approved
* There are some issues with unmounting filter swatch components on change. I haven't been able to trace the issue and fix it. __Need asssistance__

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here with the following wording: Closes #<issue> -->
Closes https://github.com/magento-research/pwa-studio/issues/361

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

Adding Layered navigation to category page.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested with Venia sample data on any product category.

Automatic tests will be added when general components and structure are approved.

## Screenshots (if appropriate):

<img width="419" alt="screen shot 2019-01-31 at 1 42 02 pm" src="https://user-images.githubusercontent.com/11479290/52054834-0a366000-255e-11e9-9fe2-e499370579c7.png">

## Proposed Labels for Change Type/Package
<!--- What types of changes does your code introduce? Let us know if this is a -->
<!--- BUG, FEATURE, DOCUMENTATION, or TEST change. -->

FEATURE

<!--- What packages are modified by this code? Let us know if this applies to -->
<!--- peregrine, pwa-buildpack, upward-js, upward-spec, venia-concept or pwa-devdocs -->

venia-concept

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.
- [x] I have linked an issue to this PR.
- [x] I have indicated the change type and relevant package(s).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All CI checks are green (linting, build/deploy, etc).
- [ ] At least one core contributor has approved this PR.

codeAdrian commented:
Reason why build is failing:

> new Octokit({headers}) is deprecated. Use {userAgent, previews} instead. See https://github.com/octokit/rest.js#client-options

sirugh commented:
@codeAdrian can you resolve the conflicts please :)

codeAdrian commented:
> @codeAdrian can you resolve the conflicts please :)

Hello @sirugh , I am out of office until Tuesday. I will fix the conflict as soon as I'm back in the office.

sirugh commented:
@codeAdrian no worries, I went ahead and resolved it. However I'm noticing that the filter is not being applied upon clicking the "Apply" button. Any ideas? We can wait till you're back next Tuesday. Enjoy your vaca!

codeAdrian commented:
Hello @sirugh , thanks for the feedback.

Regarding the applying of filters, while I was working on this feature, my understanding was that API is not ready for applying filters, so we delayed that final step in layered navigation until GraphQL API has been updated (updated to 2.3.1).

https://github.com/magento/graphql-ce/pull/259

I see that the graphQL has been updated after I submitted the PR. If the backend endpoints are updated to latest develop, I can take a look at finishing the implementation once I get back.

sirugh commented:
I've also noticed that the menu that contains the back button is missing in the menu that slides in when clicking the top-left hamburger button.

sirugh commented on packages/venia-concept/src/components/Main/main.css:
```diff
@@ -2,7 +2,6 @@
     background-color: white;
     color: rgb(var(--venia-text));
     position: relative;
-    z-index: 1;
```
This causes the main menu to appear incorrectly (without back button). Can we add this back? Was it removed intentionally?

codeAdrian replied:
This was intentional. Up until now, all modals were a child of the main container and the z-index worked as intended. 

But now the modal is part of the content which is nested deep into main container, so we had to remove the z-index on the main wrapper and add them to individual drawers. I have fixed the CSS issues and now the drawers appear correctly.

sirugh replied:
Thanks for the explanation! I did not know the history of the modals :)

sirugh commented on packages/venia-concept/src/RootComponents/Category/categoryContent.js:
```diff
@@ -21,13 +45,26 @@ class CategoryContent extends Component {
                 <div className={classes.pagination}>
                     <Pagination pageControl={pageControl} />
                 </div>
+                <FilterModal
```
I think the reason you are having z-index issues is that this `FilterModal` is a child of a category page and not of the main app like `Navigation` or `MiniCart`.

codeAdrian commented:
> I've also noticed that the menu that contains the back button is missing in the menu that slides in when clicking the top-left hamburger button.

Fixed with latest update

> Filter is not actually being applied after clicking "Apply". See below I added a filter for price but items > $100 still appear.

> Deleting all filters after applying some results in a weird scenario where you cannot "clear" them. See below I removed a single filter but the URL remains. I then click add it back but click "reset" and the url is reset.

Fixed with latest update. This was an UX issue. It is a bit unclear how the "Apply filters", "Reset Filters" and clearing individual filters from checkboxes or active filters list should interact with URL. Should it update immediately or should user apply the changes by clicking "Apply filters".

> Filter is not actually being applied after clicking "Apply". See below I added a filter for price but items > $100 still appear.

I have checked the GraphQL schema on magento 2.3-develop branch, and as far I can see, there is no way of applying attribute filters (color, size, etc.) on catalog page. Only price is supported at the moment.

https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/CatalogGraphQl/etc/schema.graphqls#L457

No info in the documentation: https://devdocs.magento.com/guides/v2.3/graphql/search-pagination.html

GraphQL playground. I have tried testing it with color filter attribute and It didn't work.

![Screen Shot 2019-03-26 at 12 25 40 PM](https://user-images.githubusercontent.com/11479290/54993560-621f9f00-4fc2-11e9-8392-04a7dbf954cb.png)

@sirugh , let me know how if the fixes work as expected and how I should proceed with this feature.

sirugh commented:
As discussed with @codeAdrian there are a few things to note here. First, is a breaking issue that Adrian will get to tomorrow. Second, the GQL endpoints _always_ return the same filters and counts. We are assuming that means they are statically defined. This may be resolved in [this issue](https://github.com/magento/graphql-ce/issues/258).

This PR is purely UX and functional in terms of the UI -- it is not possible to actually apply the filters because GQL restrictions but we did add the UX and functional work that will make them work once GQL starts accepting the filters. Do we still want to merge this? 

@jimbo @tjwiebell @jimbo @supernova-at thoughts?

codeAdrian commented:
> As discussed with @codeAdrian there are a few things to note here. First, is a breaking issue that Adrian will get to tomorrow. Second, the GQL endpoints _always_ return the same filters and counts. We are assuming that means they are statically defined. This may be resolved in [this issue](https://github.com/magento/graphql-ce/issues/258).
> 
> This PR is purely UX and functional in terms of the UI -- it is not possible to actually apply the filters because GQL restrictions but we did add the UX and functional work that will make them work once GQL starts accepting the filters. Do we still want to merge this?
> 
> @jimbo @tjwiebell @jimbo @supernova-at thoughts?

Breaking issue with pagination has been fixed in the latest update.

sirugh commented on packages/venia-concept/src/components/FilterModal/filterModal.js:
```diff
@@ -0,0 +1,112 @@
+    };
+
+    componentDidMount = () => {
+        const filterParams = this.getFilterParams(this.props.history.location);
```
I think we need to store the filters in a single location such as redux and then use that to create the chosen filters within the modal. I think the store could initialize with the values from the URL but I don't think the component should do this every time.

codeAdrian replied:
Great suggestion. Resolved in latest update. Let me know if everything works as expected.

sirugh commented on packages/venia-concept/src/index.css:
```diff
@@ -42,6 +42,10 @@ body {
     padding: 0;
 }
 
+.modalIsOpen {
```
I couldn't find this used in the code anymore. Can we delete?

codeAdrian replied:
Removed in latest update

sirugh commented on packages/venia-concept/src/RootComponents/Category/category.css:
```diff
@@ -19,6 +18,24 @@
 
+.headerButtons {
+    display: flex;
+    justify-content: center;
```
Can you add some space between the header items?

[![Image from Gyazo](https://i.gyazo.com/bc29f3b7988147a88814d482c5958e95.png)](https://gyazo.com/bc29f3b7988147a88814d482c5958e95)

codeAdrian replied:
Added space in latest update

sirugh commented on packages/venia-concept/src/components/FilterModal/FilterFooter/filterFooter.js:
```diff
@@ -0,0 +1,114 @@
+    getFooterButtons = areOptionsPristine => {
+        const { classes } = this.props;
+
+        const resetButtonClass = areOptionsPristine
```
Instead of using `getDerivedStateFromProps` we can simply check if there are any `chosenFilterOptions` when we render since the footer component is "connected" to the catalog store in redux.

codeAdrian replied:
This code remained from previous contributor. I have resolved it in latest update.

sirugh commented on packages/venia-concept/src/components/FilterModal/FilterList/filterList.js:
```diff
@@ -0,0 +1,107 @@
+                )}
+                renderItem={({ item }) => {
+                    const isActive = isFilterSelected(item);
+                    const swatchProps = {
```
Can we rename this prop to `filterProps` since not all filter items are swatches?

codeAdrian replied:
Renamed in the latest update

sirugh commented on packages/venia-concept/src/components/FilterModal/filterBlock.css:
```diff
@@ -0,0 +1,59 @@
+
+.filterList {
+    display: none;
+    padding-right: 2.75rem;
```
IMO it would be better to not apply the padding to the swatch filter, unless of course you are going off a mockup. 

[![Image from Gyazo](https://i.gyazo.com/68cc5552282f6eade7ae0350ada9a96a.png)](https://gyazo.com/68cc5552282f6eade7ae0350ada9a96a)

codeAdrian replied:
Padding is set in design: https://magento.invisionapp.com/share/S2Q41VCR45C#/screens/341558811

sirugh replied:
Hmm it just seems...off in that view. Maybe it's better on different widths. Oh well, I'm not too concerned with this, just wanted to bring it up.

jimbo replied:
The padding is fine. If I were to implement it, I would probably use the `grid-template-columns` property to make the accordion header and accordion container each have two columns, and just have the container's second column be empty, but it's not super important. Layout is achieved here.

sirugh commented on packages/venia-concept/src/components/FilterModal/filterModal.js:
```diff
@@ -0,0 +1,112 @@
+
+    render() {
+        const { classes, isModalOpen, closeModalHandler } = this.props;
+        const modalClass = isModalOpen ? classes.rootOpen : classes.root;
```
We don't need to thread this through from the root of the app. Instead, you could connect this component to the redux store and check the drawer like:

```js
mapStateToProps = ({app}) => {
 return {
    isActive: app.drawer === 'filter',
    ..otherProps
  }
})
```

Although now that I look at it, maybe @jimbo has a reason why we should connect the `RootComponents` but not the children of those components... I need to hear from him >.<

codeAdrian replied:
Fixed in latest commit

sirugh commented on packages/venia-concept/src/RootComponents/Category/category.js:
```diff
@@ -54,16 +58,18 @@ class Category extends Component {
+            idString: String(id)
+        };
+
+        const isFilterModalOpen = drawer === 'filter';
```
This is what I mean by not having to thread the prop down. Infact in the future if some other form of page besides a `category` page wants to apply some filters it will just need to use the `toggleDrawer('filter')` action and not need to pass this prop down.

codeAdrian replied:
Fixed in latest commit

sirugh commented on packages/venia-concept/src/RootComponents/Category/categoryContent.js:
```diff
@@ -21,13 +45,26 @@ class CategoryContent extends Component {
                     <div className={classes.categoryTitle}>{categoryTitle}</div>
+                    <div className={classes.headerButtons}>
+                        <button
+                            onClick={openDrawer}
```
For the same reason, could we just `mapDispatchToProps` here instead of having to thread down the handler?

codeAdrian replied:
Fixed in latest commit.

sirugh commented on packages/venia-concept/src/RootComponents/Category/categoryContent.js:
```diff
@@ -21,13 +45,26 @@ class CategoryContent extends Component {
                     <Pagination pageControl={pageControl} />
                 </div>
+                <FilterModal
+                    closeModalHandler={closeDrawer}
```
For the same reason, could we just `mapDispatchToProps` within the `FilterModal` instead of having to thread down the handler?

codeAdrian replied:
Fixed in latest commit

sirugh commented on packages/venia-concept/src/components/FilterModal/FilterList/filterSwatch.js:
```diff
@@ -0,0 +1,86 @@
+                {isActive && (
+                    <Fragment>
+                        <span
+                            className={classes.swatchLabel}
```
FWIW I feel like we could reuse a lot of the swatch component stuff here.

codeAdrian replied:
Replaced with swatch component in latest update.

sirugh commented on packages/venia-concept/src/components/FilterModal/constants.js:
```diff
@@ -0,0 +1,24 @@
+        mode: filterModes.swatch,
+        options: {
+            layout: filterLayouts.grid,
+            generateColor: true,
```
Let's remove this from the code. It's mocked anyways. We can just assume we are going to generate colors until they are provided in the response.

codeAdrian replied:
Removed "generateColor" references and config in latest update.

sirugh commented on packages/venia-concept/src/components/FilterModal/constants.js:
```diff
@@ -0,0 +1,24 @@
+        options: {
+            layout: filterLayouts.grid,
+            generateColor: true,
+            searchable: true
```
Is this another prop that could be removed when provided by the GQL? Maybe we just do away with it and assume always searchable until then.

codeAdrian replied:
I'm not sure if this is going to be returned by GraphQL out of the box. Searching through filters is not default Magento 2 functionality as far as I'm aware of. According to the design, it's only for color filters: https://magento.invisionapp.com/share/S2Q41VCR45C#/screens/341558811

sirugh replied:
Oh, I didn't even notice that this was a prop on `fashion_color`. I'm cool with this :)

sirugh commented on packages/venia-concept/src/actions/catalog/asyncActions.js:
```diff
@@ -1,4 +1,85 @@
+        const oldState = chosenFilterOptions[group] || [];
+        const newState = oldState.concat({ title, value });
+
+        dispatch(actions.filterOption.add({ newState, group }));
```
The `filterOption.add` and `filterOption.remove` functions are identical. Would it be OK to simplify into a single `filterOption.update` function that both this `filterAdd` and the below `filterRemove` function use?

codeAdrian replied:
Resolved in latest update.

sirugh commented on packages/venia-concept/src/components/FilterModal/FilterFooter/filterFooter.js:
```diff
@@ -0,0 +1,114 @@
+
+    handleApplyFilters = () => {
+        const { history, chosenFilterOptions, closeModalHandler } = this.props;
+        history.push('?' + serialize(chosenFilterOptions));
```
Can we use the `filterAdd` function here?

codeAdrian replied:
This only updated the URL and closes the modal, it doesn't affect the redux state. I'm thinking on adding "appliedFilters" redux state that will keep track of applied filters which will affect the category page listing. We'll use that function here.

codeAdrian commented on packages/venia-concept/src/reducers/catalog.js:
```diff
@@ -9,7 +9,9 @@ export const initialState = {
-    prevPageTotal: null
+    prevPageTotal: null,
+    chosenFilterOptions: {},
+    appliedFilterOptions: {}
```
You are right, thanks for catching that. It must have slipped through the previous conflicts I have resolved. Code has been updated with "prevPageTotal" references, reducers and tests removed.

sirugh replied:
Wait why was `prevPageTotal` removed?

codeAdrian commented:
> Alright! This is awesome and really good work. After getting those first few wrinkles out of the way I was able to give this a good look-see. I've left feedback on quite a few things. If you think any of it is not applicable or due to my misunderstanding (which is quite possible) please respond and let me know!
> 
> I would also like @jimbo to take a look at this PR and some of my feedback if possible. Some of the changes I requested may not fall in line with decided upon patterns that I may just not be aware of.
> 
> Cheers!

Thanks @sirugh for the detailed code review. I will go through the requested changes and resolve them one by one. I have resolved the easier ones first, so if they are okay, we can mark them as "resolved" so we can keep better track of requested changes.

sirugh commented on packages/venia-concept/src/reducers/catalog.js:
```diff
@@ -9,7 +10,8 @@ export const initialState = {
     pageSize: 6,
-    prevPageTotal: null
+    chosenFilterOptions: getFilterParams(),
+    appliedFilterOptions: getFilterParams()
```
This will get initialized but not updated. Intentional?

codeAdrian replied:
This will hold filters that are currently being applied on catalog product list. But since the API doesn't yet support it, I can remove it. Chosen filter option keep track of the filters that user edits in the modal and that will be applied to the catalog product list once user clicks "Apply filters".

codeAdrian replied:
I will remove this for now until we have a proper API support

sirugh commented on packages/venia-concept/src/RootComponents/Category/categoryContent.js:
```diff
@@ -21,13 +45,26 @@ class CategoryContent extends Component {
                     />
                     <div className={classes.categoryTitle}>{categoryTitle}</div>
+                    <div className={classes.headerButtons}>
+                        <button
```
We should conditionally render this button based on the `filters` value from data. We already return null within the modal if there are no filters, so we shouldn't even display the button in that case :)

codeAdrian replied:
Fixed in latest commit.

sirugh commented on packages/venia-concept/src/RootComponents/Category/categoryContent.js:
```diff
@@ -21,13 +45,26 @@ class CategoryContent extends Component {
                 <div className={classes.pagination}>
                     <Pagination pageControl={pageControl} />
                 </div>
+                <FilterModal
```
And maybe instead of worrying about returning null within the modal component we can conditionally render here as well?

codeAdrian replied:
Fixed in latest commit.

codeAdrian commented on packages/venia-concept/src/reducers/catalog.js:
```diff
@@ -34,14 +36,22 @@ const reducerMap = {
+    [actions.filterOption.clear]: state => {
         return {
             ...state,
-            prevPageTotal: payload
```
Sorry, my mistake. I will revert this commit. I didn't quite get which line was referenced. The add/remove diff confused me and I thought you were referring to that diff.

sirugh replied:
Ah I was referring to the `appliedFilterOptions` line because I couldn't see it used anywhere.

codeAdrian replied:
Commit has been reverted

jimbo commented on packages/venia-concept/src/components/FilterModal/filterModal.css:
```diff
@@ -0,0 +1,57 @@
+    max-width: 640px;
+    transform: translate3d(-100%, 0, 0);
+    transition-duration: 192ms;
+    transition-timing-function: cubic-bezier(0.4, 0, 1, 1);
```
I think there's a CSS var for this.

codeAdrian replied:
Fixed in latest update

jimbo commented on packages/venia-concept/src/components/FilterModal/filterBlock.js:
```diff
@@ -0,0 +1,125 @@
+            default:
+                return classes.layout;
+        }
+    };
```
Interesting approach. 👍

jimbo commented on packages/venia-concept/src/components/FilterModal/FilterSearch/filterSearch.js:
```diff
@@ -0,0 +1,88 @@
+                            <TextInput
+                                placeholder={`Search for a specific ${name}`}
+                                fieldState={{ value: filterQuery }}
+                                onChange={handleFilterSearch}
```
Consider using `onValueChange` here, so we don't have to interact with an event object or the DOM.

codeAdrian replied:
Fixed in latest commit.

jimbo commented on packages/venia-concept/src/components/FilterModal/FilterSearch/filterSearch.js:
```diff
@@ -0,0 +1,88 @@
+                item =>
+                    item.label
+                        .toUpperCase()
+                        .indexOf(filterQuery.toUpperCase()) >= 0
```
Consider using `.includes()`.

codeAdrian replied:
`includes()` is case sensitive, this approach is case insensitive. Let me know if I should go with `includes`anyway.

codeAdrian replied:
Thanks for the clarification. I have added `includes` instead of `indexOf` in the latest update.

jimbo commented on packages/venia-concept/src/components/FilterModal/FilterList/filterSwatch.js:
```diff
@@ -0,0 +1,76 @@
+                            className={classes.swatchLabel}
+                            dangerouslySetInnerHTML={{
+                                __html: label
+                            }}
```
I don't think we want to do this.

codeAdrian replied:
This component has been replaced with global swatch component in latest commit. This is not an issue anymore for Swatch component.

jimbo commented on packages/venia-concept/src/components/FilterModal/FilterList/filterDefault.js:
```diff
@@ -0,0 +1,55 @@
+                <span
+                    dangerouslySetInnerHTML={{
+                        __html: label
+                    }}
```
Why are we dangerously setting inner HTML here?

codeAdrian replied:
Some filters (like price) return HTML as their label value.

![Screen Shot 2019-03-29 at 8 31 00 AM](https://user-images.githubusercontent.com/11479290/55216451-09881600-51fd-11e9-9d1a-3a62503c6408.png)

Feel free to let me know if there is a better approach that we should use (like stripping HTML)

sirugh replied:
That's... an odd choice. @jimbo is that just an issue with our sample data or is there some practice of defining classes/html as copy at Magento?

codeAdrian replied:
Fixed in latest commit:
- Added stripHtml function
- Removed Dangerously setting HTML functions

GraphQL only returns the HTML version of the label. There is no raw string label option implemented. So this needs to be fixed with sample data or GraphQL team.

jimbo replied:
@codeAdrian Great, thank you. I'll make sure this gets brought up to the GraphQL team.

jimbo commented on packages/venia-concept/src/actions/catalog/asyncActions.js:
```diff
@@ -1,4 +1,85 @@
+        }
+    };
+
+export const filterRemove = ({ group, title, value }, history) =>
```
Minor quibble: these names stick out to me. Why not `addFilter` and `removeFilter`, which fits the usual pattern of `verbObject`?

codeAdrian replied:
Makes sense to me. Fixed in latest update.

jimbo commented on packages/venia-concept/src/actions/catalog/asyncActions.js:
```diff
@@ -1,4 +1,85 @@
 import actions from './actions';
+import mockData from './mockData';
+
+export const serialize = (params, keys = [], isArray = false) => {
```
I'll come back and review this in more depth later.

codeAdrian commented on packages/venia-concept/src/util/getFilterParamsFromUrl.js:
```diff
@@ -0,0 +1,22 @@
+
+    let urlFilterParams = {};
+
+    for (var key of params.keys()) {
```
Linter didn't warn me about this, but thanks for catching this. I have fixed this in latest commit.

sirugh replied:
Huh. Looks like we use [this config](https://github.com/magento-research/magento-eslint) for our linting. I'll open an issue regarding this :)

codeAdrian commented:
> Fails
> 🚫	
> The following unit tests did _not_ pass 😔. All tests must pass before this PR can be merged
> 
> packages/venia-concept/src/RootComponents/Category/__tests__/categoryContent.spec.js
> Generated by 🚫 [dangerJS](https://danger.systems/js) against [6922ad0](https://github.com/magento-research/pwa-studio/commit/6922ad016a50cea67277d9885ac8ec363c4ac680)

@sirugh @jimbo I might need help with fixing this test. It's a bit too complex for me.

It was caused by implementing this improvement: https://github.com/magento-research/pwa-studio/pull/797/files/468509a54281db1f1bcee18034e0c5c0d8f31a04#r269654614

sirugh commented:
> I might need help with fixing this test. 

@codeAdrian you may want to refer to other connected components to see how we test them. Another pattern I've used before is exporting the un-wrapped component as well and using that in the tests. The wrapping is just a way to provide props so if you go with that pattern you would just import that un-wrapped component within the test file and manually provide the props that would otherwise be passed by the wrapper.

We're basically doing that already, its just that all of our connected components are wrapped by a "Container" component where we do the connection. Check out the `Receipt` and `ReceiptContainer` components for an example of what I'm talking about.

codeAdrian commented:
> > I might need help with fixing this test.
> 
> @codeAdrian you may want to refer to other connected components to see how we test them. Another pattern I've used before is exporting the un-wrapped component as well and using that in the tests. The wrapping is just a way to provide props so if you go with that pattern you would just import that un-wrapped component within the test file and manually provide the props that would otherwise be passed by the wrapper.
> 
> We're basically doing that already, its just that all of our connected components are wrapped by a "Container" component where we do the connection. Check out the `Receipt` and `ReceiptContainer` components for an example of what I'm talking about.

I might go and create a wrapper around `categoryContent` component and only test the component. Great suggestion, thank you.

codeAdrian commented:
Test has been fixed as suggested by @sirugh . All tests now pass.

@jimbo let me know if I need to modify anything else.

codeAdrian commented:
Hello everyone, I have fixed the eslint issues. All tests now pass. @jimbo Any news on the reviews?

Just to let you know, I'll be out of office next week and will be unavailable.

jimbo commented:
@codeAdrian Thanks for updating the PR. I've finally got some time to take another look at this one.

codeAdrian commented:
> @codeAdrian Thanks for being patient and addressing feedback with changes. This is a good PR and we've failed to do our part to get it merged.
> 
> I do have a couple more requests, and one of them may be a bit unreasonable.
> 
> **New modals and stacking order**
> 
> I'm still concerned about adjusting all the stacking orders (`z-index`) to accommodate our first deeply nested modal. The other modals—`Navigation` and `MiniCart`—were placed as siblings to the `<main>` element, and it would be nice if we could do that with `FilterModal` as well. I think I understand why we aren't doing that yet: these filters are the category page's local concern rather than the site's global concern like the others. Still, I think it should be possible.
> 
> Can I ask you to experiment with [`createPortal`](https://reactjs.org/docs/portals.html) here? When the category mounts, try creating an element as a sibling of the other modals, capturing a `ref` to it, and rendering the FilterModal into that element. Maybe create a new branch from thsi one; if you're stuck or unhappy with how it turns out, you don't have to push anything to this branch and overwrite your work here, and we can discuss how to move forward based on how you feel about the experiment. I'm just curious because this seems like the appropriate use case for portals.
> 
> **Design**
> 
> Did @soumya-ashok provide mockups for the filter modal? I'd like to see the design more closely match the others, with the same light gray header at the top and such. Also, I'd be interested in having this modal enter from the bottom edge rather than the left edge, which would make it centered at desktop resolutions. If existing designs contradict what I'm proposing, let me know and I'll discuss with Soumya.

Hi Jimmy, thanks for the feedback. I will look into using portals. I am familiar with them but never had the chance to use them. I will look into it and let you know when I made the improvements.

Regarding the design, I am following @soumya-ashok design posted here: https://github.com/magento-research/pwa-studio/issues/361#issuecomment-456539901

Let me know if you have any more suggestions.

Regards,
Adrian

soumya-ashok commented:
@jimbo and @codeAdrian 

The filters would be requested by the shopper from a category or search results page - https://magento.invisionapp.com/share/8ZPPEYD7BGS
Having them come up from the bottom was my intention as well.

The reason they do not have the gray top navigation bar is that they are not a global element in the same way that navigation or cart is in my mind. They are contextual to a couple of instances within the shopping flow which is why both filter and sort are visually different from the main nav. or cart.

All mockups for filters - https://magento.invisionapp.com/share/F2RRNA8WMBT

codeAdrian commented:
Hello everyone, turns out that the Portal implementation was really easy. I'm really happy how it turned out. PR has been updated with:
* Added Portal for FilterModal
* Reverted changes to z-index values

Let me know if I need to fix anything else.

sirugh commented on packages/venia-concept/src/components/Modal/modal.js:
```diff
@@ -0,0 +1,6 @@
+import { createPortal } from 'react-dom';
+
+const Modal = ({ children }) =>
+    createPortal(children, document.getElementById('root'));
```
Do we care about keyboard focus for this iteration? If so we will need to do more work here.

See [here](https://reactjs.org/docs/portals.html#usage) where they discuss that it is important to manage keyboard focus with portals. And [here](https://reactjs.org/docs/accessibility.html#programmatically-managing-focus) where they have suggestions on how to proceed.

Aside from that, this is cool! Should we apply this to the other modals such as nav or the cart? cc @jimbo

codeAdrian replied:
Hello @sirugh , thank your for your feedback. I haven't seen the keyboard focus improvements being done on other drawers.

However, I have implemented the keyboard focus on the filter modal. But I'm not really happy with the implementation. For some reason, focus wouldn't be switched to modal if the `focus()` method is not in `setTimeout` function, even with the `preventDefault` function.

You can see the implementation here: `src/RootComponents/Category/categoryContent.js`.

Do you have any suggestions on how to improve this?

jimbo commented on packages/venia-concept/src/components/Modal/modal.js:
```diff
@@ -0,0 +1,6 @@
+import { createPortal } from 'react-dom';
+
+const Modal = ({ children }) =>
+    createPortal(children, document.getElementById('root'));
```
This is so simple—I love it. Two thoughts come to mind here:

- Rather than hardcoding the target to `#root`, we might consider accepting an element as a prop, with `#root` being the default
- I wonder if we can memoize the `document.getElementById()` call with `useMemo()`

codeAdrian replied:
Great suggestions. Element id prop and memoization has been added in the latest update.

jimbo replied:
I had something *slightly* different in mind for the element prop—I wanted to cover a case in which you have an element stored as a ref and don't need to query the document—so I've pushed a small change. Overall this is just what I had in mind though. 👍

jimbo commented on packages/venia-concept/src/components/FilterModal/filterModal.js:
```diff
@@ -0,0 +1,96 @@
+        mapDispatchToProps
+    ),
+    classify(defaultClasses)
+)(FilterModal);
```
Since we're calling `connect()` from `FilterModalContainer`, why are we also calling it here?

codeAdrian replied:
Good catch. I have moved the redux props to container in latest update.

dpatil-magento commented:
@jimbo @codeAdrian @sirugh    Let me know your views on below observations - If anything major - fix , minor - I will create separate issue , not needed - then we can just ignore those.

1. Select any filter and unselect - Reset Filters and Apply Filters button does not get disabled.
2. Fashion Color - If I type a color - 
      - Don't see the text in search field,
      - Filtered colors are not correct. Ex Type - Rain,   Rain and Mint 2 options are displayed.
3. Keyboard tab order focus and the way focus should move needs to be implemented.
4. If I select so many filters which does not fit in window size, then there is no way user can see the hidden filters on desktop browser. On mobile user can see by swiping left. Is it good idea to wrap in next line?
5. Say I'm on Page 4 of a category and select a filter and then click Reset filter and close filter overlay. I'm back to Page 1.
6. Filter button can move some more up, just to make UI look good.  something like below ?

<img width="345" alt="image" src="https://user-images.githubusercontent.com/40405790/57333501-99fe2400-70e2-11e9-9f15-bc3116a04146.png">

codeAdrian commented:
> @jimbo @codeAdrian @sirugh Let me know your views on below observations - If anything major - fix , minor - I will create separate issue , not needed - then we can just ignore those.
> 
> 1. Select any filter and unselect - Reset Filters and Apply Filters button does not get disabled.
> 2. Fashion Color - If I type a color -
>    
>    * Don't see the text in search field,
>    * Filtered colors are not correct. Ex Type - Rain,   Rain and Mint 2 options are displayed.
> 3. Keyboard tab order focus and the way focus should move needs to be implemented.
> 4. If I select so many filters which does not fit in window size, then there is no way user can see the hidden filters on desktop browser. On mobile user can see by swiping left. Is it good idea to wrap in next line?
> 5. Say I'm on Page 4 of a category and select a filter and then click Reset filter and close filter overlay. I'm back to Page 1.
> 6. Filter button can move some more up, just to make UI look good.  something like below ?

Hi @dpatil-magento , thank you for your feedback. Let me address each point separately.

1. Thanks for catching that. This issue has been fixed.
2. Color search issue has been fixed.
3. Jimmy has commented on this topic here https://github.com/magento-research/pwa-studio/pull/797#pullrequestreview-234262932. I suggest creating a new issue for this because it also includes navigation, minicart, checkout and account modals.
4. I have enabled a scrollbar to render as browser (or OS) default. Meaning that desktop will render it as a standard scrollbar, and mobile will render it in their own styled scrollbar (hidden or transparent). This ensures equal UX across all devices.
5. I am assuming if the user changes filters, the catalog item listing will change and user would like to start browsing from the start so they can view items they might have missed. Let me know if there is a UX reason that we might want to keep the current page when changing filters.
6. Styles have been adjusted in the latest update.

soumya-ashok commented:
@codeAdrian and @dpatil-magento - I need a little more context on 5. since I am unable to view the deployment.

5. If a shopper is browsing a category, and at some point decides to apply a filter, it makes sense to me that they are back at "Page 1" because they are in a new view within the category now. I think this is what you're saying as well Adrian?

4. I am not sure how much space was around the filter button but my intent was to actually provide a touch area that would be comfortable for someone with larger hands. Keeping active elements too close together can lead to accidental clicks. That being said we should not be adjusting UI padding / spacing one-off. I would like @jimbo to weigh in on this as well.

dpatil-magento commented:
@codeAdrian   I discussed with @soumya-ashok and 5 also needs be fixed. If the filters are engaged on page 3, but no filters are actually applied, when the filter screen is closed, the shopper would still be on the same page they engaged the filters from.

@jimbo @codeAdrian  Per discussion with @soumya-ashok - here are some more changes needed. Let Me know if we need to address as part of this ticket or create new ticket?
A. Filter screen is designed to be a full screen overlay
B. Category filter should not be displayed when user is already on Category page. It make sense to display Category when user is search results page.
C. The filter pill text size could be knocked down a point
D. Filters should be category aware. i.e  If I select a category then clothing styles filter option should be updated.
E.  Selected filter pills (header area on filter screen) be sticky so they can stay in the shopper's view?

jimbo commented:
1. Good suggestion. Conditionally disabling modal actions is the right idea. Good fix.
2. Good suggestion. Color search works now, but throws when clearing the field.
3. Good suggestion, but not now. Keyboard support should be addressed as a new issue.
4. It's fine to let the filter tag container scroll horizontally. No fix necessary.
5. Fair suggestion. Pagination should only change when the *applied filters* change, since that's when the data changes. If it overcomplicates the code, though, let me know and we can waive this requirement.
6. @dpatil-magento Let's reserve the minor style feedback for Soumya. 😅 It looks fine as-is, and we certainly don't want *less* spacing. I'll work with @soumya-ashok on how to adjust spacing and sizing for this page, but it won't block this PR.

jimbo commented:
A. The bottom sheet approach used here is pretty solid. I wouldn't make any changes to how the filter modal is presented here.
B. I don't have an opinion on this.
C. This is another minor style thing, but I disagree. We don't use `15px` text anywhere in Venia (other than text fields and only for Safari compatibility reasons), and `16px` is our most-preferred size. If anything, the tags would grow larger in order to be valid touch areas, but I think they're acceptable as-is.
D. Yes, it'd be nice if filters were category-aware, but I don't think we have that ability right now. I assume we simply get a flat list of potential filters and filter values from GraphQL? @codeAdrian
E. You mean within the modal? It's fine if we fix the height of the modal and allow the modal body to overflow with a vertical scrollbar, but I'd like to avoid any new sticky toolbars.

codeAdrian commented:
Hello everyone, this PR has been updated with the following fixes:
* Fixed the "Clear filter search" issue. This now works as intended.
* Fixed the page reset issue on applying and removing filters. The easiest way of fixing this was just to apply this fix for all cases.

To answer the comments:
* Regarding the category GraphQL query. There is an GraphQL issue which returns the same filters for any category or search query. So I would recommend to wait for GraphQL fix before making any adjustments to filters groups.
* Regarding the modal, it looks better when it covers a part of the bottom section of the screen. Also the transition when opening and closing looks much better when it doesn't cover the entire screen.
* @soumya-ashok can you provide me with the list of style adjustments I need to make, taking into account @jimbo and @dpatil-magento suggestions?

dpatil-magento commented:
@codeAdrian   Fixes on Category page looks good.  Some observations on **Search Result page** -

7. "Fashion Size" and "Price" filters are missing.
8. In Filter overlay I select a filter and click on Reset Filters button -  
      a- Filter Overlay closes with shadowed window instead of just clearing out the filters
      b- User is taken back to Home page instead of keeping them on Search page with results.
9. In Filter overlay I select a filter and click on Apply Filters button then user taken back to Home page

soumya-ashok commented:
@jimbo Per the design, the filter pill text size is actually smaller than body copy. We can keep the touch area the same but I think the text is too large relative to the size of the pill. Personal opinion.

On sticking the top of the filter modal - I think the visual treatment on the checkbox changing when selected is enough reinforcement for the shopper to know that their action has succeeded so we can scroll this area without too much negative impact to the user. I was originally thinking that if they had multiple filters applied it might be good to keep that within their view but they could just scroll to the top to see it if they really wanted to.

@codeAdrian 
With having the filters be full screen here are my thoughts - We have multiple accordions which collapse and expand within the area that already taken up vertical space and we cannot predict how many filters there will be within each accordion since this depends entirely on how the store is set up. Why would we not use the full height of the device and allow the user to see more filters at a time? I do not see an advantage in having this be a modal of restricted height.

codeAdrian commented:
> "Fashion Size" and "Price" filters are missing.
@dpatil-magento This is a Backend / GraphQL issue. We only show filters that the graphQL query returns.

> I do not see an advantage in having this be a modal of restricted height.
@soumya-ashok Modal height is not restricted. It can expand to the full height once more accordions are open. You can see how that looks right how. I can adjust it to make it full height tomorrow if it's going to look better.

codeAdrian commented:
Hello everyone, following fixes have been applied in the new update:
* Fixes for search results filtering issues
* Fix for spacing above Filter button

Let me know if you find anything else for me to fix and I'll fix it on Monday. If there is something that can be fixed sooner than that, feel free to push a fix to the PR.

dpatil-magento commented:
@codeAdrian   Fixes looks good.  One more scenario does not look correct - 
10- Apply a filter. Go back to filter overlay and uncheck the applied filter.   Url still has the filter in it and Apply/Reset Filter button is disabled.   There is no way user can remove the filter now.

Note - The above scenario works fine if I remove filter by clicking the  filter in filter header but again this breaks the flow when say user engage filter overlay for the first time and selects 2 filters and removes one filter via filter header. Check the URL remaining filter is already applied even though user has not clicked Apply filters button.

Is it good idea to make filters to take effects only when Reset Filters and Apply Filters buttons are clicked? 

@soumya-ashok  Please provide your inputs on above.

codeAdrian commented:
Hello everyone,
PR has been updated with the fix for the following scenario:
> Apply a filter. Go back to filter overlay and uncheck (via check box) the applied filter. Url still has the filter in it and Apply/Reset Filter button is disabled. There is no way user can remove the filter now.

Let me know if anything else needs fixing.

dpatil-magento commented:
@codeAdrian  Recent fix looks good.  Please check following scenario's :
11:  Apply filter on Search page. Now navigate to category page and go to filter overlay.  The previously applied filters on search page still show up on Category page filter.
12: On Filter overlay select filters and close filter overlay by clicking X (Close) icon or outside overlay area to close filter overlay without applying selected filters. Now Go back to filter overlay, previously selected filters are still in checked state.

codeAdrian commented:
Hello @dpatil-magento , I have updated the PR with the fixes you have reported. I have went through the all edge cases and scenarios I can think of and I haven't noticed anything out of place. Let me know if you find anything out of place.

dpatil-magento commented:
Thanks @codeAdrian all fixes looks good.

@jimbo @soumya-ashok  Please let me know if we need to create ticket for below which is pending from GraphQL just for tracking purpose. And also if you have any other suggestions for this ticket. 

D: Filters should be category aware

soumya-ashok commented:
@codeAdrian - Not sure if you implemented the full height on the modal for the mobile context, but I'm still seeing the old treatment. The modal height varying with the expanding accordions is a bit jarring. We would deal with filters differently on desktop, so I won't get into that here.

@dpatil-magento - I didn't understand your question exactly but unchecking the filters removes the tags within the filter modal. However, if I were to close the modal and reopen it, those old filters I removed are still there.

Applying filters isn't doing anything at the moment. I don't see a results page.

Also, when I select a color, the swatch expands to a rectangle. This should not be happening. Screenshot below.

![Screen Shot 2019-05-16 at 2 44 52 PM](https://user-images.githubusercontent.com/25960021/57882421-4332d180-77e9-11e9-8b24-255d93b6cf24.png)

codeAdrian commented:
Hello @soumya-ashok thank you for feedback.

Regarding the filters not applying after changing, I'm not sure if the attribute filters input has been added to GraphQL backend. @jimbo , @sirugh feel free to let me know if I'm in the wrong. I've been checking out the Catalog GraphQL schema and didn't see the attribute filter input: https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/CatalogGraphQl/etc/schema.graphqls#L279

Regarding the style issues (full height filters modal and swatch width), they have been fixed in the latest update. Let me know if there is anything else that needs to be fixed.

dpatil-magento commented:
Thanks @codeAdrian fixes looks good.

sirugh commented:
<!--
  1 failure:  node` failed.
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  node` failed.
  </td>
    </tr>
  </tbody>
</table>



### Log


```sh
Error:  Error: Could not find a `dangerfile.js` or `dangerfile.ts` in the current working directory.
    at Object.dangerfilePath (/codebuild/output/src451119981/src/github.com/magento-research/pwa-studio/node_modules/danger/distribution/commands/utils/fileUtils.js:26:11)
    at Object.<anonymous> (/codebuild/output/src451119981/src/github.com/magento-research/pwa-studio/node_modules/danger/distribution/commands/danger-runner.js:92:42)
    at step (/codebuild/output/src451119981/src/github.com/magento-research/pwa-studio/node_modules/danger/distribution/commands/danger-runner.js:33:23)
    at Object.next (/codebuild/output/src451119981/src/github.com/magento-research/pwa-studio/node_modules/danger/distribution/commands/danger-runner.js:14:53)
    at fulfilled (/codebuild/output/src451119981/src/github.com/magento-research/pwa-studio/node_modules/danger/distribution/commands/danger-runner.js:5:58)
    at process._tickCallback (internal/process/next_tick.js:68:7)

```

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0e122d302bf0cdfc01db36461ece854019a87e93
</p>

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento-research/pwa-studio?pullRequest=797) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-research/pwa-studio?pullRequest=797) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Starotitorov<br/>:x: Adrian Bece<hr/>**Adrian Bece** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

jimbo commented:
@codeAdrian I've resolved the conflicts that were created when we merged #1211, which converted `Category` to a functional component using hooks. Take a look and see if you have any feedback; otherwise we'll try to merge this Monday.

codeAdrian commented:
@jimbo Thanks for resolving the conflicts. I really appreciate it. I have fixed a minor issue and cleaned up a commented-out piece of code that was leftover from the merge. I am happy to proceed with the merge on Monday. Cheers!

dpatil-magento commented:
Did another round of QA, all looks good. @jimbo - Good to merge.
